pkg/pgmodel/lreader/labels_reader.go

```diff
@@ -78,9 +79,49 @@ func (lr *labelsReader) LabelValues(labelName string) ([]string, error) {
 	return labelValues, nil
 }
 
+const getLabelNamesBelongingToTenantNames = `
+SELECT ARRAY_AGG(a.key) FROM
+(
+	SELECT 
+	  key 
+	FROM 
+	  _prom_catalog.label WHERE id IN (
+		SELECT 
+		  UNNEST(labels) as valid_label_ids 
+		FROM 
+		  "_prom_catalog".series 
+		WHERE 
+		  labels && (SELECT array_agg(id :: INTEGER) FROM _prom_catalog.label WHERE key = '__tenant__' AND value = ANY($1)) -- List of valid tenant names goes here.
+	  ) GROUP BY 1 ORDER BY 1
+) a
+`
+
 // LabelNames implements the LabelReader interface. It returns all distinct
 // label names available in the database.
 func (lr *labelsReader) LabelNames() ([]string, error) {
+	if lr.authConfig != nil {
+		// Multi-tenancy is enabled. Hence, we have to use a different LabelNames() query
+		// only when some tenants are authorized, i.e., when using NewSelectiveTenancyConfig().
+		//
+		// Note: Label names of non-tenants will not be sent. Only label names belonging to
+		// authorized tenants will be sent. This means, if -metrics.multi-tenancy.allow-non-tenants
+		// is enabled, the non-tenant labels will not be answered ATM.
+		validTenants, allTenantsValid := lr.authConfig.ValidTenants()
+		if len(validTenants) == 0 {
+			return []string{}, nil
```

Comment:
Maybe log a warning here that no valid tenants are found. It would make it easier for debugging if no labels are retutned